src/isl/util.rs

```diff
@@ -27,6 +29,224 @@ pub enum Range {
+                        unreachable!("Lower range boundary value must not be max")
+                    }
+                    (_, Min) => {
+                        unreachable!("Upper range boundary value must not be min")
+                    }
+                }
+            }
+            Range::Decimal(start, end) => {
+                match (start, end) {
+                    (Min, Max) => {
+                        unreachable!("Range boundaries can not be min and max together (i.e. range::[min, max] is not allowed)")
+                    },
+                    (Min, Value(end_value, boundary_type)) => {
+                        match end_value {
+                            Decimal(decimal_end_value) => {
+                                Ok(decimal_end_value > &value.as_decimal().unwrap() || (*boundary_type == RangeBoundaryType::Inclusive && decimal_end_value == value.as_decimal().unwrap()))
+                            }
+                            _ => unreachable!("Decimal range can only have decimals as lower and upper range boundary value")
+                        }
+                    }
+                    (Value(start_value, boundary_type), Max) => {
+                        match start_value {
+                            Decimal(decimal_start_value) => {
+                                Ok(decimal_start_value < &value.as_decimal().unwrap() || (*boundary_type == RangeBoundaryType::Inclusive && decimal_start_value == value.as_decimal().unwrap()))
+                            }
+                            _ => unreachable!("Decimal range can only have decimals as lower and upper range boundary value")
+                        }
+                    },
+                    (Value(start_value, start_boundary_type), Value(end_value, end_boundary_type)) => {
+                        match (start_value, end_value) {
+                            (Decimal(range_start_value), Decimal(range_end_value)) => {
+                                let range_start_validation = range_start_value < &value.as_decimal().unwrap() || (*start_boundary_type == RangeBoundaryType::Inclusive && range_start_value == value.as_decimal().unwrap());
+                                let range_end_validation = range_end_value > &value.as_decimal().unwrap() || (*end_boundary_type == RangeBoundaryType::Inclusive && range_end_value == value.as_decimal().unwrap());
+                                Ok(range_start_validation && range_end_validation)
+                            },
+                            _ => unreachable!("Decimal range can only have decimals as lower and upper range boundary value")
+                        }
+                    }
+                    (Max, _) => {
+                        unreachable!("Lower range boundary value must not be max")
+                    }
+                    (_, Min) => {
+                        unreachable!("Upper range boundary value must not be min")
+                    }
+                }
+            }
```

Comment:
suggestion—The comparisons for the `(Min, Value)`, `(Value, Value)`, and `(Value, Max)` cases have repeated code. It would eliminate some of the repeated code if you refactored this so that the value we're testing is unwrapped once at the beginning of the block and so that upper and lower range bounds are considered separately (like below).
```rust
Range::Decimal(start, end) => {
    let unwrapped_value = &value.as_decimal().unwrap();
    let is_in_lower_bound = match start {
        Min => true,
        Value(start_value, boundary_type) => match start_value {
            Decimal(min_value) => match boundary_type {
                RangeBoundaryType::Inclusive => min_value <= unwrapped_value,
                RangeBoundaryType::Exclusive => min_value < unwrapped_value,
            },
            _ => unreachable!("Decimal range can only have decimals as lower and upper range boundary value"),
        },
        Max => unreachable!("Cannot have 'Max' as the lower range boundary")
    };
    // let is_in_upper_bound = match end {
    // ...
    is_in_upper_bound && is_in_lower_bound
}
```